### ISSUE TITLE
`needless_borrow` suggestion causes error when removing `ref` in match

### ISSUE THREAD

**Issue by 5225225** [rust-lang/rust-clippy#9049]
### Summary

I was running clippy on rustc and ended up with this error

### Lint Name

needless_borrow

### Reproducer

I tried this code:

```rust
pub struct Impl<'a> {generics: &'a Generics<'a> }
pub struct Generics<'a>(&'a String);

pub enum ItemKind<'hir> {
    Impl(&'hir Impl<'hir>),
    Fn(&'hir Generics<'hir>),
}

impl ItemKind<'_> {
    pub fn generics(&self) -> &Generics<'_> {
        match *self {
            ItemKind::Fn(ref generics)
            | ItemKind::Impl(Impl { ref generics }) => generics,
        }
    }
}

fn main() {}
```

I saw this happen:

```
warning: this pattern creates a reference to a reference
  --> src/main.rs:12:26
   |
12 |             ItemKind::Fn(ref generics)
   |                          ^^^^^^^^^^^^
13 |             | ItemKind::Impl(Impl { ref generics }) => generics,
   |                                     ^^^^^^^^^^^^
   |
   = note: `#[warn(clippy::needless_borrow)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow
help: try this
   |
12 ~             ItemKind::Fn(generics)
13 ~             | ItemKind::Impl(Impl { generics }) => generics,
   |
```

I expected to see this happen:

Nothing. You can't remove the `ref` here without causing a compile error. (Or, suggest the correct thing to do here, I'm not too familiar with `ref` patterns so there might be a correct thing here.)

### Version

```text
rustc 1.63.0-nightly (fdca237d5 2022-06-24)
binary: rustc
commit-hash: fdca237d5194bf8a1c9b437ebd2114d1c2ba6195
commit-date: 2022-06-24
host: x86_64-unknown-linux-gnu
release: 1.63.0-nightly
LLVM version: 14.0.5
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by Jarcho** [comment]
Either matching on `self` or using `ItemKind::Impl(&Impl { generics })` the second pattern would work. The suggestion should include one of those.

**Comment by kraktus** [comment]
Strangely (to my eyes), this work too:
```rust
impl ItemKind<'_> {
    pub fn generics(&self) -> &Generics<'_> {
        match *self {
            ItemKind::Fn(generics) => generics,
            ItemKind::Impl(Impl { generics }) => generics,
        }
    }
}
```
Which I'd have assumed is the same as an `Or` match
